core/src/main/scala/com/evolutiongaming/kafka/flow/ConsumerFlow.scala

```diff
@@ -132,4 +104,50 @@ object ConsumerFlow {
 
   }
 
+  private class RebalanceListenerImpl[F[_]: BracketThrowable: LogOf](
+    consumer: Consumer[F],
+    flows: Map[Topic, TopicFlow[F]]
+  ) extends RebalanceListener[F] {
+
+    def onPartitionsAssigned(topicPartitions: NonEmptySet[TopicPartition]) =
+      foreachTopic(topicPartitions) { (topic, flow, partitions) =>
+        for {
+          log <- log[F]
+          _ <- log.prefixed(topic).info(s"$partitions assigned")
+          partitions <- partitions.toNonEmptyList traverse { partition =>
+            consumer.position(TopicPartition(topic, partition)) map (partition -> _)
+          }
+          _ <- log.prefixed(topic).info(s"committed offsets: $partitions")
+          _ <- flow.add(partitions.toNes)
+        } yield ()
+      }
+
+    def onPartitionsRevoked(topicPartitions: NonEmptySet[TopicPartition]) =
+      foreachTopic(topicPartitions) { (topic, flow, partitions) =>
+        for {
+          log <- log[F]
+          _ <- log.prefixed(topic).info(s"$partitions revoked, removing from topic flow")
+          _ <- flow.remove(partitions)
+        } yield ()
+      }
+
+    def onPartitionsLost(topicPartitions: NonEmptySet[TopicPartition]) =
+      foreachTopic(topicPartitions) { (topic, flow, partitions) =>
+        for {
+          log <- log[F]
+          _ <- log.prefixed(topic).info(s"$partitions lost, removing from topic flow")
+          _ <- flow.remove(partitions)
+        } yield ()
+      }
+
+    private def foreachTopic[A](topicPartitions: NonEmptySet[TopicPartition])(
+      f: (Topic, TopicFlow[F], NonEmptySet[Partition]) => F[A]
```

Comment:
This signature looks a bit heavyweight... It also adds a cognitive load because `foreachTopic` is kind of a new DSL. :thinking: 

Should we make a method out of `topicPartitions.filter(_.topic == topic).map(_.partition).toNes` instead, so we can still call something like `traverse_` on a result?

Or I am nitpicking here?

Reply:
That's relevant complain, fixed